libs/ml/src/core/node.cpp

```diff
@@ -0,0 +1,236 @@
+}
+
+template <typename TensorType>
+void Node<TensorType>::SetBatchOutputShape(const Shape &new_shape)
+{
+  op_ptr_->SetBatchOutputShape(new_shape);
+}
+
+template <typename TensorType>
+void Node<TensorType>::SetBatchInputShapes(const ShapeVector &new_shapes)
+{
+  op_ptr_->SetBatchInputShapes(new_shapes);
+}
+
+template <typename TensorType>
+const ShapeVector &Node<TensorType>::BatchInputShapes()
+{
+  return op_ptr_->BatchInputShapes();
+}
+
+/**
+ * @brief BatchOutputShape computes an output shape of the Node, if only 1 data slice (e.g.
+ * with batch size == 1) is provided to the Graph input. If there is no cached output shape,
+ * the method is recursively called until either a cached shape or input Node is encountered.
+ * @return vector of SizeType.
+ */
+template <typename TensorType>
+Shape Node<TensorType>::BatchOutputShape()
+{
+  Shape const &candidate = op_ptr_->BatchOutputShape();
+  bool const   op_is_subgraph =
+      (std::dynamic_pointer_cast<std::shared_ptr<SubGraph<TensorType>>>(op_ptr_) != nullptr);
+
+  // If the underlying Op is simple (e.g. not a SubGraph) and its shape is already known (cached)
+  // this shape can be returned immediately.
+  if (!candidate.empty() && !op_is_subgraph)
+  {
+    if (OperationType() == OpType::OP_PLACEHOLDER)
+    {
+      FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached a placeholder input node : " +
+                                        this->name_ + " " + OutputShapeAsString(candidate));
+    }
+    if (OperationType() == OpType::OP_WEIGHTS)
+    {
+      FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached weights node : " + this->name_ + " " +
+                                        OutputShapeAsString(candidate));
+    }
+    return candidate;
+  }
+
+  // If there is no input nodes, shape deduction can not go further.
+  if (input_nodes_.empty())
+  {
+    if (candidate.empty())
+    {
+      // If there is no input nodes, but underlying Op's shape is not known - the Graph
+      // probably is incorrect.
+      FETCH_LOG_ERROR(name_.c_str(),
+                      "Shape deduction reached a Graph leaf with empty shape " + this->name_);
+    }
+    return candidate;
+  }
+
+  ShapeVector input_shapes;
+  for (auto const &i : input_nodes_)
+  {
+    auto input_node_ptr = i.lock();
+    if (!input_node_ptr)
+    {
+      throw std::runtime_error("Unable to lock weak pointer.");
+    }
+
+    // If there are valid input nodes, make a deeper recursive call to each of the previous nodes.
+    Shape const in_shape = input_node_ptr->BatchOutputShape();
+
+    if (in_shape.empty())
+    {
+      if (input_node_ptr->OperationType() != OpType::OP_PLACEHOLDER)
+      {
+        FETCH_LOG_INFO(name_.c_str(),
+                       "Got an empty shape as return from non-placeholder layer! : " +
+                           input_node_ptr->GetNodeName());
+      }
+      continue;
+    }
+
+    input_shapes.emplace_back(in_shape);
```

Comment:
i think it's a little clearer to formulate it like this:

if (!in_shape.empty())
{
  input_shapes.emplace_back(in_shape);
}
else
{
  // error handling...
}

Reply:
In this method I've tried to keep the next approach: if something went wrong, we return, else we continue. This way each next instruction is executed only if there were no errors before. Keeping an "else" block will result in less clearer flow here, as it is a bit harder to understand which of the blocks ("if" or "else" one) is for good case.